### PR TITLE
pkg/cmd: support cache sizes in percentages of free memory

### PR DESCRIPTION
**Pull request by jzelinskie**
This PR has been reworked a bunch, but now adds support for cache max cost flags to use percentages of free memory in addition to bytes.

Lots of other cleanup was done in the process.

### CODE REVIEW

File: pkg/cmd/server/cacheconfig.go [hunk 1]
```diff
@@ -58,8 +58,8 @@ func (cc *CacheConfig) Complete() (*cache.Config, error) {
 // RegisterCacheConfigFlags registers flags for a ristretto-based cache.
 func RegisterCacheConfigFlags(flags *pflag.FlagSet, config *CacheConfig, flagPrefix string) {
 	flagPrefix = stringz.DefaultEmpty(flagPrefix, "cache")
-	flags.StringVar(&config.MaxCost, flagPrefix+"-max-cost", defaultMaxCost, "the maximum cost to be stored in the cache, in bytes")
-	flags.Int64Var(&config.NumCounters, flagPrefix+"-num-counters", defaultNumCounters, "the number of keys to track")
+	flags.Int64Var(&config.MaxEntries, flagPrefix+"-max-entries", defaultMaxEntries, "maximum number of cache entries")
```
Comment (josephschorr) [inline pkg/cmd/server/cacheconfig.go:61]: Maybe we should have the old flag be a hidden and if specified, have it warn or fail?

File: pkg/cmd/server/cacheconfig.go [hunk 2]
```diff
@@ -2,64 +2,79 @@ package server
...
+func RegisterCacheFlags(flags *pflag.FlagSet, flagPrefix string, config, defaults *CacheConfig) {
 	flagPrefix = stringz.DefaultEmpty(flagPrefix, "cache")
-	flags.StringVar(&config.MaxCost, flagPrefix+"-max-cost", defaultMaxCost, "the maximum cost to be stored in the cache, in bytes")
-	flags.Int64Var(&config.NumCounters, flagPrefix+"-num-counters", defaultNumCounters, "the number of keys to track")
-	flags.BoolVar(&config.Metrics, flagPrefix+"-metrics", false, "whether metrics should be maintained for the cache. WARNING: Incurs a performance penality.")
-	flags.BoolVar(&config.Disabled, flagPrefix+"-disabled", false, "if true, fully disables the cache. WARNING: Incurs a performance penality.")
+	flags.StringVar(&config.MaxCost, flagPrefix+"-max-cost", defaults.MaxCost, "upper bound cache size in bytes or percent of available memory")
+	flags.Int64Var(&config.NumCounters, flagPrefix+"-num-counters", defaults.NumCounters, "number of TinyLFU samples to track")
+	flags.BoolVar(&config.Metrics, flagPrefix+"-metrics", false, "enable cache metrics")
+	flags.BoolVar(&config.Enabled, flagPrefix+"-enabled", true, "enable caching")
```
Comment (vroldanbet) [inline pkg/cmd/server/cacheconfig.go:79]: There two parameters are ignoring the provided default:

```suggestion
	flags.BoolVar(&config.Metrics, flagPrefix+"-metrics", defaults.Metrics, "enable cache metrics")
	flags.BoolVar(&config.Enabled, flagPrefix+"-enabled", defaults.Enabled, "enable caching")
```

File: pkg/cmd/serve.go [hunk 3]
```diff
@@ -26,12 +26,17 @@ func RegisterServeFlags(cmd *cobra.Command, config *server.Config) {
...
 	if err := cmd.Flags().MarkHidden("ns-cache-expiration"); err != nil {
 		panic("failed to mark flag hidden: " + err.Error())
 	}
-	server.RegisterCacheConfigFlags(cmd.Flags(), &config.NamespaceCacheConfig, "ns-cache")
+	server.RegisterCacheFlags(cmd.Flags(), "ns-cache", &config.NamespaceCacheConfig, &server.CacheConfig{
+		Enabled:     true,
+		Metrics:     false,
+		NumCounters: 1_000,
+		MaxCost:     "16MiB",
+	})
```
Comment (vroldanbet) [inline pkg/cmd/serve.go:39]: it wasn't immediately clear this wasn't a hardcoded configuration but a default and that CLI args take precedence. Perhaps extracting these into a variable named `namespaceCacheDefaults` would make it clearer?

File: pkg/cmd/serve.go [hunk 4]
```diff
@@ -57,8 +62,18 @@ func RegisterServeFlags(cmd *cobra.Command, config *server.Config) {
...
+		Metrics:     false,
+		NumCounters: 10_000,
+		MaxCost:     "30%",
+	})
+	server.RegisterCacheFlags(cmd.Flags(), "dispatch-cluster-cache", &config.ClusterDispatchCacheConfig, &server.CacheConfig{
+		Enabled:     true,
+		Metrics:     false,
+		NumCounters: 100_000,
+		MaxCost:     "70%",
+	})
```
Comment (vroldanbet) [inline pkg/cmd/serve.go:76]: I remember discussing we could have this cache be shared between "dispatch cluster cache" and "cluster cache". Is that something that will be addressed in a follow up?

File: pkg/cmd/server/cacheconfig.go [hunk 5]
```diff
@@ -2,64 +2,79 @@ package server
...
+	parsedPercent, err := strconv.ParseUint(percent, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse percentage: %w", err)
+	}
+
+	if parsedPercent > 100 {
+		return 0, fmt.Errorf("percentage greater than 100")
+	}
+
+	return freeMemory / 100 * parsedPercent, nil
```
Comment (vroldanbet) [inline pkg/cmd/server/cacheconfig.go:69]: can we please add some tests to `parsePercent` 🙏🏻

File: pkg/cmd/server/cacheconfig.go [hunk 6]
```diff
@@ -2,64 +2,79 @@ package server
...
+		maxCost uint64
+		err     error
+	)
 
-	maxCost, err := humanize.ParseBytes(cc.MaxCost)
+	if strings.HasSuffix(cc.MaxCost, "%") {
+		maxCost, err = parsePercent(cc.MaxCost)
+	} else {
+		maxCost, err = humanize.ParseBytes(cc.MaxCost)
+	}
```
Comment (vroldanbet) [inline pkg/cmd/server/cacheconfig.go:46]: may be worth testing this as well

**Review by vroldanbet — COMMENTED** [review]
Nice clean up! 👏🏻 Found a defect with defaults, some missing tests and potential readability improvements. I'm also wondering how this refactor would help if we decide to have dispatch cache and cluster dispatch cache be shared.

File: pkg/cmd/serve.go (hunk 4 above)
Comment (jzelinskie) [reply pkg/cmd/serve.go]: We're putting that out of scope for now. I've downscoped this to the original goal which is to let folks specify percentages of available memory.

**Review by vroldanbet — APPROVED** [review]

**Pull request merged**